### PR TITLE
feat: implement border highlight

### PR DESCRIPTION
**Pull request by romgrk**
Link: https://github.com/MunifTanjim/nui.nvim/pull/142

### CODE REVIEW

File: lua/nui/popup/border.lua [hunk 1]
```diff
@@ -561,6 +567,29 @@ function Border:set_text(edge, text, align)
   line:render(self.bufnr, self.popup.ns_id, linenr)
 end
 
+---@param border_highlight string For example, "WarningMsg"
+function Border:set_highlight(border_highlight)
```
Comment (MunifTanjim) [inline lua/nui/popup/border.lua:571]: This method should also have support for `internal.type == "simple"`.

**Review by MunifTanjim — DISMISSED** [review]
LGTM 👍🏼

**Comment by MunifTanjim** [comment]
@romgrk Would you be fixing the stylua error? Otherwise I can go ahead and push the fix?

**Comment by romgrk** [comment]
I have stylued the PR and rebase on main, shoud be good to go once the checks complete.

File: lua/nui/text/init.lua [hunk 2]
```diff
@@ -35,6 +35,19 @@ function Text:set(content, extmark)
...
+  if extmark then
+    self.extmark = is_type("string", extmark) and { hl_group = extmark } or vim.deepcopy(extmark)
+    self.extmark.id = nil
+  else
+    self.extmark = { id = nil, hl_group = nil }
+  end
+
+  return self
+end
+
```
Comment (MunifTanjim) [inline lua/nui/text/init.lua:50]: I don't remember it being here last time I reviewed. 🤔

Is this a related change? Why is this needed?

Comment (romgrk) [reply lua/nui/text/init.lua:50]: It was there previously. For complex borders, the `chars` is an array of `Text` that we want to change the highlight. I exposed `Text:set_highlight` rather than accessing internal state.

Comment (MunifTanjim) [reply lua/nui/text/init.lua:50]: Isn't it the same as `text:set(text:content(), highlight)`?

Comment (MunifTanjim) [reply lua/nui/text/init.lua:50]: If we change the `Text:init` a little bit, we can even do `text = Text(text, new_hl)`

File: lua/nui/popup/border.lua [hunk 3]
```diff
@@ -547,6 +567,34 @@ function Border:set_text(edge, text, align)
...
+    return
+  end
+
+  internal.winhighlight = merge_winhl(self.popup._.win_options.winhighlight, "FloatBorder", border_highlight)
+
+  internal.highlight = nil
+  internal.highlight = normalize_highlight(internal)
+
+  for _, item in pairs(internal.char) do
+    item:set_highlight(internal.highlight)
```
Comment (MunifTanjim) [inline lua/nui/popup/border.lua:588]:
```suggestion
    item:set(item:content(), internal.highlight)
```

File: tests/nui/menu/init_spec.lua [hunk 4]
```diff
@@ -45,6 +45,22 @@ describe("nui.menu", function()
...
+          Menu.item("a"),
+        },
+      })
+
+      menu:mount()
+
+      h.assert_buf_lines(menu.bufnr, {
+        "a",
+      })
+    end)
```
Comment (MunifTanjim) [inline tests/nui/menu/init_spec.lua:63]: I think this got added back when rebasing. This can safely be deleted.

```suggestion
```

File: lua/nui/text/init.lua (hunk 2 above)
Comment (romgrk) [reply lua/nui/text/init.lua:50]: I feel like adding a `set_highlight` is preferrable due to its increased readbility.

Regarding `text:set(text:content(), new_hl)`: it has the disadvantage that half the characters aren't related to the action in question in this use-case, in addition to the unneeded argument passing.
Regarding `text = Text(text:content(), new_hl)`: I don't like the unneeded memory allocation; the fact that luajit is fast and has a proper GC doesn't mean we shouldn't help it, in particular in what is meant to be a core library.

Comment (MunifTanjim) [reply lua/nui/text/init.lua:50]: Well, this particular use-case is an internal one. And doing `text:set(text:content(), new_hl)` in this PR is not really doing any harm.

In future if we see that changing only the highlight for an existing text is a very common use-case, we can add that method later.

We don't have to do it right now, since it's not really necessary for this PR.

**Review by MunifTanjim — APPROVED** [review]

**Comment by MunifTanjim** [comment]
Thanks you! 🎉
